presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineRemotePartitionedExchangeEncoding.java

```diff
@@ -0,0 +1,136 @@
+import com.google.common.annotations.VisibleForTesting;
+
+import static com.facebook.presto.SystemSessionProperties.getMinColumnarEncodingChannelsToPreferRowWiseEncoding;
+import static com.facebook.presto.spi.plan.ExchangeEncoding.ROW_WISE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE_STREAMING;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.sql.planner.plan.Patterns.Exchange.scope;
+import static com.facebook.presto.sql.planner.plan.Patterns.Exchange.type;
+import static com.facebook.presto.sql.planner.plan.Patterns.exchange;
+
+public class DetermineRemotePartitionedExchangeEncoding
+        implements Rule<ExchangeNode>
+{
+    private static final Pattern<ExchangeNode> PATTERN = exchange()
+            .with(scope().equalTo(REMOTE_STREAMING))
+            .with(type().equalTo(REPARTITION));
+
+    private final boolean nativeExecution;
+    private final boolean prestoSparkExecutionEnvironment;
+
+    public DetermineRemotePartitionedExchangeEncoding(boolean nativeExecution, boolean prestoSparkExecutionEnvironment)
+    {
+        this.nativeExecution = nativeExecution;
+        this.prestoSparkExecutionEnvironment = prestoSparkExecutionEnvironment;
+    }
+
+    @Override
+    public Pattern<ExchangeNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return nativeExecution || prestoSparkExecutionEnvironment;
+    }
+
+    @Override
+    public Result apply(ExchangeNode node, Captures captures, Context context)
+    {
+        if (prestoSparkExecutionEnvironment) {
+            // In Presto on Spark, row-wise encoding is always used for non-special shuffles (i.e., excluding broadcast, single, and arbitrary shuffles).
+            // To accurately reflect this in the plan, the exchange encoding is set here.
+            return determineForPrestoOnSpark(node);
+        }
+        if (nativeExecution) {
+            return determineForNativeExecution(context.getSession(), node);
+        }
+        // Presto Java runtime does not support row-wise encoding
+        return Result.empty();
+    }
+
+    private Result determineForPrestoOnSpark(ExchangeNode node)
+    {
+        // keep columnar for special cases
+        if (node.getPartitioningScheme().isSingleOrBroadcastOrArbitrary()) {
+            return Result.empty();
+        }
+        if (node.getPartitioningScheme().getEncoding() == ROW_WISE) {
+            // leave untouched if already visited
+            return Result.empty();
+        }
+        // otherwise switch to row-wise
+        return Result.ofPlanNode(node.withRowWiseEncoding());
+    }
+
+    private Result determineForNativeExecution(Session session, ExchangeNode node)
+    {
+        // keep columnar for special cases
+        if (node.getPartitioningScheme().isSingleOrBroadcastOrArbitrary()) {
```

Comment:
I'm curious why we check that it's not single or broadcast or arbitrary instead of checking that it is a fixed hash distribution? What other things are we trying to include here?

Reply:
Technically a connector can provide any partitioning function. In most cases it is hash based, but it can be anything. The engine provides a special case shortcuts for certain partitioning functions it knows the semantics of.

For example the engine knows that for "single", "arbitrary" or "broadcast" partitioning can be done at the page level and doesn't necessarily need to be done at the row level. This allows the engine to avoid additional cost associated with splitting pages. Hence these partitioning functions are rather an exception than a rule, hence it felt it makes more sense to single them out.